crmsh/migration.py

```diff
@@ -350,29 +345,69 @@ def _check_saphana_resource_agent(handler: CheckResultHandler, resource_agents:
+
+    def __init__(self):
+        self._unsupported = dict()
+        for line in pkgutil.get_data(
+                'crmsh', 'migration-unsupported-resource-agents.txt'
+        ).decode('ascii').splitlines():
+            parts = line.split(',', 3)
+            parts.extend(itertools.repeat('', 2))
+            unsupported = self.__resource_agent_from_str(parts[0])
+            alternative = parts[1]
+            is_deprecated = parts[2] == "deprecated"
+            self._unsupported[unsupported] = self.UnsupportedState(
+                self.__resource_agent_from_str(alternative) if alternative != '' else None,
+                is_deprecated,
+            )
+
+    def get_unsupported_state(self, resource_agent):
+        return self._unsupported.get(resource_agent)
+
+    @staticmethod
+    def __resource_agent_from_str(s: str):
+        parts = s.split(':', 3)
         m_class = parts[0]
         m_provider = parts[1] if len(parts) == 3 else None
         m_type = parts[-1]
-        ret.add(cibquery.ResourceAgent(m_class, m_provider, m_type))
-    return ret
-
+        return cibquery.ResourceAgent(m_class, m_provider, m_type)
 
 
 def _check_removed_resource_agents(
         handler: CheckResultHandler,
-        supported_resource_agents: typing.Set[cibquery.ResourceAgent],
+        agent_type_message: str,
+        unsupported_resource_agents: UnsupportedResourceAgentDetector,
         resource_agents: typing.Iterable[cibquery.ResourceAgent],
 ):
-    unsupported_resource_agents = [x for x in resource_agents if x not in supported_resource_agents]
-    if unsupported_resource_agents:
-        handler.handle_problem(False, '', [
-            '* ' + ':'.join(x for x in resource_agent if x is not None) for resource_agent in unsupported_resource_agents
+    unsupported: typing.List[typing.Tuple[cibquery.ResourceAgent, UnsupportedResourceAgentDetector.UnsupportedState]] = list()
+    deprecated: typing.List[typing.Tuple[cibquery.ResourceAgent, UnsupportedResourceAgentDetector.UnsupportedState]] = list()
+    for x in resource_agents:
+        unsupported_state = unsupported_resource_agents.get_unsupported_state(x)
+        if unsupported_state.is_deprecated:
+            deprecated.append((x, unsupported_state))
+        else:
+            unsupported.append((x, unsupported_state))
+    if unsupported:
+        handler.handle_problem(False, f'The following {agent_type_message} will be removed in SLES 16.', [
+            '* {}{}'.format(
+                ':'.join(x for x in resource_agent if x is not None),
+                ': please replace it with {}'.format(
+                    ':'.join(x for x in unsupported_state.alternative if x is not None)
+                ) if unsupported_state.alternative is not None else ''
+            )
+            for resource_agent, unsupported_state in unsupported
+        ])
+    if deprecated:
+        handler.handle_problem(False, f'The following {agent_type_message} will be deprecated in SLES 16.', [
```

Comment:
"deprecated" is not "FAIL". It should be "WARN". It is not a migration-stopper.

```
[WARN] The following resource agents will be deprecated in SLES 16.
       * ocf:heartbeat:Raid1: please replace it with ocf:heartbeat:mdraid
```

Reply:
Now we have many dimensions: _removed_ vs _deprecated_, _stopper_ vs _non-stopper_, fix-_before_ vs _after_-migration. For example, external/sbd is _removed_ + _non-stopper + fix _after_-migration. We need to add more columns to the data file.